sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaSimpleWritableDataSource.java

```diff
@@ -0,0 +1,374 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package test.org.apache.spark.sql.connector;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.*;
+import org.apache.spark.deploy.SparkHadoopUtil;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.connector.TestingV2Source;
+import org.apache.spark.sql.connector.catalog.SessionConfigSupport;
+import org.apache.spark.sql.connector.catalog.SupportsWrite;
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+import org.apache.spark.sql.connector.read.ScanBuilder;
+import org.apache.spark.sql.connector.write.*;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.apache.spark.util.SerializableConfiguration;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.Iterator;
+
+public class JavaSimpleWritableDataSource implements TestingV2Source, SessionConfigSupport {
```

Comment:
It would be great if you add a class description here, @kevincmchen .

Reply:
Do we need to use `TestingV2Source`? Shall we use `SimpleTableProvider` instead?

Reply:
Yea I think we should follow the scala version and use `SimpleTableProvider`. `TestingV2Source.inferSchema` doesn't match the schema here.

Reply:
> Yea I think we should follow the scala version and use `SimpleTableProvider`. `TestingV2Source.inferSchema` doesn't match the schema here.

using SimpleTableProvider will cause  compile error and the error is 
`Error:(53, 8) java: test.org.apache.spark.sql.connector.JavaSimpleWritableDataSource is not abstract and does not override abstract method org$apache$spark$sql$internal$connector$SimpleTableProvider$$loadedTable_$eq(org.apache.spark.sql.connector.catalog.Table) in org.apache.spark.sql.internal.connector.SimpleTableProvider`

the reason is that scala traits cannot be extended in Java if they have any implementation details.

Reply:
so i have 2 options:
1. create a java implementation of SimpleTableProvider
2. use TestingV2Source and override inferSchema function.

Do you think which one is better?

Reply:
Ah, I see. Makes sense.

Maybe we can change the schema of this writable v2 source (as well as the read part), to match `TestingV2Source`, so that the scala version can also use `TestingV2Source`, to be consistent with the java version.